### PR TITLE
refactor(core): extract persist_large_output helper and annotate truncated output

### PR DESCRIPTION
**Pull request by chaliy**
## What
Extract shared `persist_large_output()` helper and `annotate_truncated_output()` from the inline hook logic in `tool_output_persistence.rs`. When exec output exceeds the 16 KiB budget, the helper persists stdout/stderr as separate files under `/.outputs/`, and the annotation appends a file reference to truncated stdout so agents know they can `read_file` with offset/limit.

## Why
The persistence logic in `PersistOutputHook` was inline and not reusable. Extracting it as a `pub` helper allows other sandbox tools or hooks to persist large output using the same logic. The truncation annotation (EVE-245) gives agents an explicit pointer to the full output file.

Note: the core persistence behavior (separate `.stdout`/`.stderr` files, `output_files` array, `/.outputs/` paths) was already merged via #1197. This PR adds the **helper extraction** and **truncation annotation** on top.

## How
- Extracted `persist_large_output()` as a pub async helper with `PersistResult` return type
- Added `annotate_truncated_output()` to append file reference to truncated text
- Refactored `PersistOutputHook` to delegate to the helper instead of inline logic
- Hook now annotates truncated stdout with `[full output saved to /workspace{path} — use read_file with offset/limit]`
- Updated `EXEC_OUTPUT_HINT` to mention `output_files` array
- Added 7 unit tests for `persist_large_output()` covering thresholds, path sanitization, line count, edge cases

## Risk
- Low
- Pure refactor of existing persistence logic into a reusable helper
- No new trust boundaries, no API surface changes, no migrations
- Annotation is additive — appended to already-truncated stdout text

## Security
- Threat categories reviewed: TM-FS, TM-DOS, TM-LLM
- TM-FS: `safe_id` sanitization preserved (alphanumeric + dash + underscore only). No user-controlled path components. Test verifies path traversal chars are stripped.
- TM-DOS: Persistence bounded by existing exec timeouts. No new unbounded allocations.
- TM-LLM: Static annotation format string. No injection surface.

## Checklist
- [x] Tests added or updated
- [x] Backward compatibility considered
- [x] Security review performed against relevant threat model categories
- [x] All review comments addressed (code change or written reasoning)

### CODE REVIEW

**Comment by chaliy** [comment]
## Review feedback addressed

Rebased onto latest `main` (resolving merge conflicts) and addressed all 7 Copilot review comments:

1. **Fixed Unicode artifacts** — replaced `��` replacement characters with `—` in header comment and `annotate_truncated_output` format string
2. **Updated stale design comment** — now correctly references `/.outputs/{safe_id}.stdout` and `.stderr` (not the old `/.exec-logs/{id}.log`)
3. **Fixed PersistResult doc comment** — describes paths as "session VFS form" (e.g. `/.outputs/{safe_id}.stdout`), not "/workspace-prefixed"
4. **Fixed guard comment** — now reads "Only persist if stdout exceeds its budget or stderr exceeds 4096 bytes"
5. **Clarified `full_output` semantic** — added inline comment: "full_output points to the stdout file only; stderr (if persisted) is available via the output_files array"
6. **Added unit tests for `persist_large_output()`** — 7 new async tests covering: below-budget returns None, stdout above budget, stderr above threshold, both persisted, ID sanitization (path traversal), empty ID rejection, line count accuracy

Note: the core persistence behavior (separate `.stdout`/`.stderr` files, `output_files` array) was already merged via #1197. This PR's remaining value is the **extracted `persist_large_output()` pub helper** and **`annotate_truncated_output()`** — making the persistence logic reusable and annotating truncated output with file references for the agent.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capabilities/tool_output_persistence.rs [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
 //
 // Persists full exec tool output to session VFS before truncation,
 // enabling lossless retrieval via read_file/grep. The LLM gets a
-// truncated summary with a `full_output` path to the full log.
+// truncated summary with a `full_output` path and `output_files`
+// array pointing to the full log.
 //
 // Design decisions:
-// - Implemented as PostToolExecHook, not baked into each tool — VFS
+// - Implemented as PostToolExecHook, not baked into each tool �� VFS
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/tool_output_persistence.rs]: Fixed — replaced Unicode replacement characters with proper em dash (`—`) in the header comment.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capabilities/tool_output_persistence.rs [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 //   persistence is a cross-cutting concern the tool shouldn't know about
 // - Reads `persist_output` hint from ToolDefinition to decide what to persist
 // - Writes to /.exec-logs/{tool_call_id}.log (dot-prefixed, session-scoped)
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/tool_output_persistence.rs]: Fixed — updated design comment to reference `/.outputs/{safe_id}.stdout` and `.stderr` (the paths used after rebasing onto main which moved from `/.exec-logs/` to `/.outputs/`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capabilities/tool_output_persistence.rs [hunk 3]
```diff
@@ -19,8 +22,93 @@ use serde_json::json;
...
 use crate::tool_types::{ToolCall, ToolDefinition, ToolResult};
-use crate::traits::ToolContext;
+use crate::traits::{SessionFileStore, ToolContext};
+use crate::typed_id::SessionId;
+
+/// Result of persisting large exec output to session VFS.
+pub struct PersistResult {
+    /// Path to the persisted stdout file (in /workspace-prefixed form for LLM).
+    pub stdout_path: Option<String>,
+    /// Path to the persisted stderr file, if non-empty stderr was persisted.
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/tool_output_persistence.rs]: Fixed — doc now reads "Path to the persisted stdout file in session VFS form (e.g. `/.outputs/{safe_id}.stdout`)" — no mention of `/workspace` prefix since that's added by the caller.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capabilities/tool_output_persistence.rs [hunk 4]
```diff
@@ -19,8 +22,93 @@ use serde_json::json;
...
+///
+/// This is the shared helper that any sandbox tool or hook can call.
+pub async fn persist_large_output(
+    file_store: &Arc<dyn SessionFileStore>,
+    session_id: SessionId,
+    tool_call_id: &str,
+    stdout: &str,
+    stderr: &str,
+) -> Option<PersistResult> {
+    // Only persist if stdout exceeds budget
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/tool_output_persistence.rs]: Fixed — comment now reads "Only persist if stdout exceeds its budget or stderr exceeds 4096 bytes".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capabilities/tool_output_persistence.rs [hunk 5]
```diff
@@ -19,8 +22,93 @@ use serde_json::json;
...
+    } else {
+        None
+    }
+}
+
+/// Build the annotated truncated stdout string with file reference.
+pub fn annotate_truncated_output(truncated: &str, file_path: &str, full_size: usize) -> String {
+    let size_kb = full_size / 1024;
+    format!(
+        "{truncated}\n\n[full output saved to /workspace{file_path} ({size_kb} KiB) �� use read_file with offset/limit]"
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/tool_output_persistence.rs]: Fixed — replaced Unicode replacement characters with proper em dash (`—`) in the annotation format string.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capabilities/tool_output_persistence.rs [hunk 6]
```diff
@@ -87,42 +175,62 @@ impl PostToolExecHook for PersistOutputHook {
...
+                        let annotated =
+                            annotate_truncated_output(current_stdout, path, stdout.len());
+                        obj.insert("stdout".to_string(), json!(annotated));
+                    }
+                    output_files.push(format!("/workspace{path}"));
+                    obj.insert("full_output".to_string(), json!(path));
+                    obj.insert(
+                        "total_lines".to_string(),
+                        json!(persist_result.stdout_total_lines),
+                    );
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/tool_output_persistence.rs:214]: Good catch. Went with option (b): added an inline comment clarifying "`full_output` points to the stdout file only; stderr (if persisted) is available via the `output_files` array". This matches the existing contract on main (PR #1197 already set `full_output` to stdout-only) — the separation is intentional since agents typically care about stdout for re-reading and stderr for diagnostics.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capabilities/tool_output_persistence.rs [hunk 7]
```diff
@@ -19,8 +22,93 @@ use serde_json::json;
...
+    file_store: &Arc<dyn SessionFileStore>,
+    session_id: SessionId,
+    tool_call_id: &str,
+    stdout: &str,
+    stderr: &str,
+) -> Option<PersistResult> {
+    // Only persist if stdout exceeds budget
+    if stdout.len() <= EXEC_OUTPUT_BUDGET && stderr.len() <= 4096 {
+        return None;
+    }
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/tool_output_persistence.rs:60]: Added 7 async unit tests with a lightweight `MockFileStore`:
- `test_persist_large_output_returns_none_below_budget`
- `test_persist_large_output_persists_stdout_above_budget`
- `test_persist_large_output_persists_stderr_above_threshold`
- `test_persist_large_output_both_stdout_and_stderr`
- `test_persist_large_output_sanitizes_id` (path traversal)
- `test_persist_large_output_empty_id_returns_none`
- `test_persist_large_output_line_count`

---
_Generated by [Claude Code](https://claude.ai/code)_
